lib/ansible/modules/packaging/os/portage.py

```diff
@@ -472,7 +472,7 @@ def main():
             newuse=dict(default=False, type='bool'),
             changed_use=dict(default=False, type='bool'),
             oneshot=dict(default=False, type='bool'),
-            noreplace=dict(default=False, type='bool'),
+            noreplace=dict(default=True, type='bool'),
```

Comment:
Changing the default would break the module for people relying on the current behavior.

Reply:
@s-hertel can you confirm this causes breakage of existing behavior? Currently noreplace is broken, it does not work and has no effect either way. I cannot believe that anyone is relying on that broken behavior.

The main issue I can see here is for packaging updating, trying to update an installed package, and not being able to because of noreplace being set to true. I need to test that out to confirm. If it does not effect package upgrades, then there would be no change from current behavior. If it does, I can modify the code such that if upgrading, remove the noreplace flag or set it to no/false.

You have to set ```noreplace: no/false``` for the behavior to change with this modification. Otherwise it should function the same either way.

Reply:
@s-hertel I have tested both ```state: latest``` and ```update: yes``` without having anything set for ```noreplace```, meaning it is using the new default of true ```noreplace=dict(default=True, type='bool')```, and still performing the package update.

I am not able to replicate any breakage with this change. If nobody else can show how this change does break the module. Then, any breakage is theoretical and not factual.

Reply:
By break, we mean that the result is different than current expectations and that breaks the contract with current users,  not that it causes an actual code failure.

Reply:
@bcoca thanks, that sort of makes sense. However, I am confused with current expectations. This is a broken feature, a broken code aspect of the module. I have a hard time believing anyone is expecting this broken behavior. I think no one has experienced it, thus it was never reported as an issue. This really does not change anything from the user perspective.

This is fixing an actual broken feature, broken in code. I am not clear what these contractual expectations are from users.

Also for future reference, is there an actual contract or rules to follow regarding such? I am curious about this contract with users aspect. Thanks!

Reply:
I'm not that sure it is a broken feature, my expectation is closer to what you expressed, but i still don't see this as broken .. at most, misleading.

As for actual contract, no, nothing singed, it is an understanding that things don't change arbitrarily unless they are clearly a bug, to be honest, this is on the line for me.

Reply:
- It is broken in the sense that there is no way to re-emerge/re-install an installed package, nor can such be forced. It simply cannot be done.
- It is broken because the flag has no effect, you can set noreplace to either yes/true or no/false and it has no effect.

This is clearly a bug, one that has existed for a very long time. I can only assume people are not seeking to re-install an existing package often if ever. Only doing initial installs and upgrades via ```update: true/yes``` and/or ```state: latest``` If anyone did, they would quickly find out they cannot, and this issue would have come up a long time ago.

If you look at the original issue #55365 report testing expected behavior, it failed on several. It was already not meeting expectations. Thus, this change to fix the behavior to meet expectations and provide an actual way to re-emerge/re-install an existing package.

Also I set this to True in the module to meet current expectations. That way if someone is just checking to see say ```state: present``` or ```state: installed``` it would not change the behavior at all. I spent considerable time on this trying to figure out how to fix the issue without changing how the module functions in other uses.

The only other way is to introduce another variable, and have two that control if a package is merged or not. That did not seem ideal, required more code, and likely more chance of issue. Not to mention confusion to the end user, as they see two options for noreplace rather than one.

Reply:
Given this is a portage module for Gentoo, where software is built against what is installed. It is expected behavior to be able to re-emerge/re-install an existing package. You may need to in order to fix breakage from linking. Typically done via revdep-rebuild, but there can be various scenarios where a package is needed to be re-emerge/re-installed, that is not an upgrade.

Presently there is no way to re-install an existing package. This is the default behavior of portage. ```emerge busybox``` Will emerge busybox no matter if it is installed or not.

This is presently impossible with the portage module. You can install it if not present, or upgrade/update it to the latest version. But that is all you can do. You can not re-emerge the package, despite it being the default for portage normally. This module prevents such entirely.

This PR and fix allows the module to function like portage. It allows the user to control if a package is re-emerge/re-installed or not, regardless if it is present or not. This change is the only way that one can perform such task via the portage module.

Reply:
@wltjr consider that most users of this plugin deal with pre built binaries, it is rare to handle any number of Gentoo machines and have them all rebuild packages themselves (distcc cluster aside).

I'm a Gentoo user also, so for my workstation, what you say makes all the sense in the world, i'm trying to balance the other perspective, of those that manage a fleet that install pre-built binaries. That said, I do lean towards your point of this being a bug. 

To clarify, I had not read your original post on my first response, I was trying to clarify what the previous poster meant by 'breaking'.

Reply:
@bcoca I also deal with pre-build binaries on my many systems, as anyone with a large number of Gentoo systems should. That is another scenario where you may need to re-emerge a package, if you re-built it on the build server and linked it to different stuff. You need to push out the new binary to other servers, or have them re-emerge the binary from the binhost.

I am that other scenario you are describing. I maintain this module mostly for servers not workstations, though I do use ansible to keep my workstations maintained as well. That is why I wanted to test to ensure that this did not change the other uses. I believe I have tested most all scenarios. Thus far, I have yet to see anything negative or unwanted from this change. Given how broken it was before, I do not think the change will have any impact unless someone uses it, sets ```noreplace: no/false```. That is the only time anything will happen, and it will be a wanted effect if they are setting that option. But even if that is never set, it does not seem to change anything. I tried to test all scenarios.

No worries on the break clarification, I understand the differences on breaking from code or other perspectives. I believe I tried to address all and this change does not cause any breakage in code, function, or unwanted behavior. It simply fixes a broken feature, makes an option work that had no effect before.

Reply:
Anything left to review or change here? Or can we proceed? Thanks!